bakery/codec.go

```diff
@@ -23,109 +28,158 @@ type caveatId struct {
+	case 'e':
+		// 'e' will be the first byte if the caveatid is a base64 encoded JSON object.
+		return decodeJSONCaveatId(key, id)
+	default:
+		return caveatInfo{}, errgo.Newf("caveat id has unsupported version %d", id[0])
 	}
 }
 
-func (d *boxDecoder) decodeCaveatId(id string) (rootKey []byte, condition string, err error) {
-	data, err := base64.StdEncoding.DecodeString(id)
+// decodeJSONCaveatId attempts to decode a base64 encoded JSON id. This
+// encoding is nominally version -1.
+func decodeJSONCaveatId(key *KeyPair, id []byte) (caveatInfo, error) {
+	data := make([]byte, (3*len(id)+3)/4)
+	n, err := base64.StdEncoding.Decode(data, id)
 	if err != nil {
-		return nil, "", fmt.Errorf("cannot base64-decode caveat id: %v", err)
+		return caveatInfo{}, errgo.Notef(err, "cannot base64-decode caveat id")
 	}
+	data = data[:n]
 	var tpid caveatId
 	if err := json.Unmarshal(data, &tpid); err != nil {
-		return nil, "", fmt.Errorf("cannot unmarshal caveat id %q: %v", data, err)
+		return caveatInfo{}, errgo.Notef(err, "cannot unmarshal caveat id %q", data)
 	}
-	var recordData []byte
-
-	recordData, err = d.encryptedCaveatId(tpid)
+	if !bytes.Equal(key.Public.Key[:], tpid.ThirdPartyPublicKey.Key[:]) {
+		return caveatInfo{}, errgo.New("public key mismatch")
+	}
+	if tpid.FirstPartyPublicKey == nil {
+		return caveatInfo{}, errgo.New("target service public key not specified")
+	}
+	// The encrypted string is base64 encoded in the JSON representation.
+	secret, err := base64.StdEncoding.DecodeString(tpid.Id)
 	if err != nil {
-		return nil, "", err
+		return caveatInfo{}, errgo.Notef(err, "cannot base64-decode encrypted data")
 	}
-	var record caveatIdRecord
-	if err := json.Unmarshal(recordData, &record); err != nil {
-		return nil, "", fmt.Errorf("cannot decode third party caveat record: %v", err)
+	var nonce [NonceLen]byte
+	if copy(nonce[:], tpid.Nonce) < NonceLen {
+		return caveatInfo{}, errgo.Newf("nonce too short %x", tpid.Nonce)
 	}
-	return record.RootKey, record.Condition, nil
+	cid, err := key.open(tpid.FirstPartyPublicKey, &nonce, secret)
+	if err != nil {
+		return caveatInfo{}, errgo.Notef(err, "cannot decrypt caveat id %#v", tpid)
+	}
+	var record caveatIdRecord
+	if err := json.Unmarshal(cid, &record); err != nil {
+		return caveatInfo{}, errgo.Notef(err, "cannot decode third party caveat record")
+	}
+	// Do not include the Peer Public Key as it was not returned
+	// before and might be used by the caller to determine a legacy
+	// encoding was used. Perhaps caveatInfo should contain an
+	// explicit version instead?
+	return caveatInfo{
+		RootKey:   record.RootKey,
+		Predicate: record.Condition,
+	}, nil
 }
 
-func (d *boxDecoder) encryptedCaveatId(id caveatId) ([]byte, error) {
-	if d.key == nil {
-		return nil, fmt.Errorf("no public key for caveat id decryption")
+// decodeCaveatIdV0 decodes a version 0 caveat id.
+func decodeCaveatIdV0(key *KeyPair, id []byte) (caveatInfo, error) {
+	if len(id) < v0EncryptedPartOffset {
```

Comment:
Again, I think the offsets don't make things much clearer here.

I think I'd suggest something like:

	if len(id) < 1+ 4+KeyLen+NonceLen+box.Overhead {
     		return errgo.New("caveat id too short")
     	}
 
     	myKeyPrefix := id[0:4]
     	if !bytes.Equal(key.Public[0:4], myKeyPrefix) {
     		...
     	}
     	id = id[4:]
     
     	var peerPublicKey PublicKey
     	copy(peerPublicKey[:], id)
     	id = id[KeyLen:]
     	
     	var nonce [NonceLen]byte
     	copy(nonce[:], id)
     	id = id[NonceLen:]
     	
     	data, err := key.open(&peerPublicKey, &nonce, id)